pkg/cover/canonicalizer.go

```diff
@@ -0,0 +1,130 @@
+	for _, module := range modules {
+		canonicalModules[module.Name] = uint32(module.Addr)
+	}
+
+	// Store sorted canonical address keys.
+	canonicalModuleKeys := make([]uint32, len(modules))
+	setModuleKeys(canonicalModuleKeys, modules)
+	return &Canonicalizer{
+		modules:    canonicalModules,
+		moduleKeys: canonicalModuleKeys,
+	}
+}
+
+func (can *Canonicalizer) NewInstance(modules []host.KernelModule) *CanonicalizerInstance {
+	// Save sorted list of module offsets.
+	moduleKeys := make([]uint32, len(modules))
+	setModuleKeys(moduleKeys, modules)
+
+	// Create a hash between the "canonical" module addresses and each VM instance.
+	instToCanonicalMap := make(map[uint32]*canonicalizerModule)
+	canonicalToInstMap := make(map[uint32]*canonicalizerModule)
+	for _, module := range modules {
+		canonicalAddr := can.modules[module.Name]
+		instAddr := uint32(module.Addr)
+
+		canonicalModule := &canonicalizerModule{
+			offset:  int(instAddr) - int(canonicalAddr),
+			endAddr: uint32(module.Size) + canonicalAddr,
+		}
+		canonicalToInstMap[canonicalAddr] = canonicalModule
+
+		instModule := &canonicalizerModule{
+			offset:  int(canonicalAddr) - int(instAddr),
+			endAddr: uint32(module.Size) + instAddr,
+		}
+		instToCanonicalMap[instAddr] = instModule
+	}
+
+	return &CanonicalizerInstance{
+		canonical:          *can,
+		moduleKeys:         moduleKeys,
+		instToCanonicalMap: instToCanonicalMap,
+		canonicalToInstMap: canonicalToInstMap,
+	}
+}
+
+func (ci *CanonicalizerInstance) Canonicalize(cov []uint32) {
+	convertModulePCs(ci.moduleKeys, ci.instToCanonicalMap, cov)
+}
+
+func (ci *CanonicalizerInstance) Decanonicalize(cov []uint32) {
+	convertModulePCs(ci.canonical.moduleKeys, ci.canonicalToInstMap, cov)
+}
+
+// Store sorted list of addresses. Used to binary search when converting PCs.
+func setModuleKeys(moduleKeys []uint32, modules []host.KernelModule) {
+	// Determine the address that defines the end of the modules.
+	// Any pc over this value will be considered to be part of the main kernel.
+	for idx, module := range modules {
+		// Truncate PCs to uint32, assuming that they fit into 32 bits.
+		// True for x86_64 and arm64 without KASLR.
+		moduleKeys[idx] = uint32(module.Addr)
+	}
+
+	// Sort modules by address.
+	sort.Slice(moduleKeys, func(i, j int) bool { return moduleKeys[i] < moduleKeys[j] })
+}
+
+func convertModulePCs(moduleKeys []uint32, conversionHash map[uint32]*canonicalizerModule, cov []uint32) {
+	// Skip conversion if modules are not used.
+	if len(moduleKeys) == 0 {
+		return
+	}
+	for idx, pc := range cov {
+		// Determine which module each pc belongs to.
+		i, _ := sort.Find(len(moduleKeys), func(i int) int {
+			if pc < moduleKeys[i] {
+				return -1
+			}
+			return +1
+		})
+		// Check if address is above the first module offset.
+		if i > 0 {
+			module := conversionHash[moduleKeys[i-1]]
+			// If the address is within the found module add the offset.
+			if pc < module.endAddr {
```

Comment:
I might be wrong, but if `moduleKeys` are actually the module start addresses in a non-decreasing order, you might want to find the first `i` such that `pc >= moduleKeys[i]`, then verify `pc < module[i].endAddr`. The corner case is that `i == len(moduleKeys)`.

Reply:
Sorry I may be misunderstanding your comment, but I think that's what this does. `Find()` (https://pkg.go.dev/sort#Find) finds the smallest index for `i` such that `pc >= moduleKeys[i]`.

However, as `i` is >= the index, it is one above the necessary module. Because of this, the corner case is then that `i == 0`, or `i` is less than the first module.

Once I've verified the corner case, I verify that `pc <  module[i - 1]`

Would it help if I clarified/modified the Find function?

Reply:
I attempted to clarify this in the code. Let me know if I need to make any other changes!

Reply:
I got it now, thanks. You're looking for the first module that starts after the `pc`. It's all correct then.